src/IMG_jpg.c

```diff
@@ -343,89 +343,103 @@ static void output_no_message(j_common_ptr cinfo)
-#endif
-    if (setjmp(jerr.escape)) {
+    vars->cinfo.err = lib.jpeg_std_error(&vars->jerr.errmgr);
+    vars->jerr.errmgr.error_exit = my_error_exit;
+    vars->jerr.errmgr.output_message = output_no_message;
+    if (setjmp(vars->jerr.escape)) {
         /* If we get here, libjpeg found an error */
-        lib.jpeg_destroy_decompress(&cinfo);
-        if ( surface != NULL ) {
-            SDL_DestroySurface(surface);
+        lib.jpeg_destroy_decompress(&vars->cinfo);
+        if (vars->surface) {
+            SDL_DestroySurface(vars->surface);
+            vars->surface = NULL;
         }
-        SDL_SeekIO(src, start, SDL_IO_SEEK_SET);
         SDL_SetError("JPEG loading error");
-        return NULL;
+        return;
     }
 
-    lib.jpeg_create_decompress(&cinfo);
-    jpeg_SDL_IO_src(&cinfo, src);
-    lib.jpeg_read_header(&cinfo, TRUE);
+    lib.jpeg_create_decompress(&vars->cinfo);
+    jpeg_SDL_IO_src(&vars->cinfo, src);
+    lib.jpeg_read_header(&vars->cinfo, TRUE);
 
-    if (cinfo.num_components == 4) {
+    if (vars->cinfo.num_components == 4) {
         /* Set 32-bit Raw output */
-        cinfo.out_color_space = JCS_CMYK;
-        cinfo.quantize_colors = FALSE;
-        lib.jpeg_calc_output_dimensions(&cinfo);
+        vars->cinfo.out_color_space = JCS_CMYK;
+        vars->cinfo.quantize_colors = FALSE;
+        lib.jpeg_calc_output_dimensions(&vars->cinfo);
 
         /* Allocate an output surface to hold the image */
-        surface = SDL_CreateSurface(cinfo.output_width, cinfo.output_height, SDL_PIXELFORMAT_BGRA32);
+        vars->surface = SDL_CreateSurface(vars->cinfo.output_width, vars->cinfo.output_height, SDL_PIXELFORMAT_BGRA32);
     } else {
         /* Set 24-bit RGB output */
-        cinfo.out_color_space = JCS_RGB;
-        cinfo.quantize_colors = FALSE;
+        vars->cinfo.out_color_space = JCS_RGB;
+        vars->cinfo.quantize_colors = FALSE;
 #ifdef FAST_JPEG
-        cinfo.scale_num   = 1;
-        cinfo.scale_denom = 1;
-        cinfo.dct_method = JDCT_FASTEST;
-        cinfo.do_fancy_upsampling = FALSE;
+        vars->cinfo.scale_num   = 1;
+        vars->cinfo.scale_denom = 1;
+        vars->cinfo.dct_method = JDCT_FASTEST;
+        vars->cinfo.do_fancy_upsampling = FALSE;
 #endif
-        lib.jpeg_calc_output_dimensions(&cinfo);
+        lib.jpeg_calc_output_dimensions(&vars->cinfo);
 
         /* Allocate an output surface to hold the image */
-        surface = SDL_CreateSurface(cinfo.output_width, cinfo.output_height, SDL_PIXELFORMAT_RGB24);
+        vars->surface = SDL_CreateSurface(vars->cinfo.output_width, vars->cinfo.output_height, SDL_PIXELFORMAT_RGB24);
     }
 
-    if ( surface == NULL ) {
-        lib.jpeg_destroy_decompress(&cinfo);
-        SDL_SeekIO(src, start, SDL_IO_SEEK_SET);
+    if (!vars->surface) {
+        lib.jpeg_destroy_decompress(&vars->cinfo);
         SDL_SetError("Out of memory");
```

Comment:
Not a regression, but on failure didn't `SDL_CreateSurface()` already set the error indicator to something more specific?

Reply:
Possibly, but all other loaders do the same, so I didn't change the behavior in this patch.

Maybe a subject for another patch.